agent/structs/structs_test.go

```diff
@@ -152,6 +152,82 @@ func testServiceNode() *ServiceNode {
 	}
 }
 
+func TestNode_IsSame(t *testing.T) {
+	id := types.NodeID("e62f3b31-9284-4e26-ab14-2a59dea85b55")
+	node := "mynode1"
+	address := ""
+	datacenter := "dc1"
+	n := &Node{
+		ID:              id,
+		Node:            node,
+		Datacenter:      datacenter,
+		Address:         address,
+		TaggedAddresses: make(map[string]string),
+		Meta:            make(map[string]string),
+		RaftIndex: RaftIndex{
+			CreateIndex: 1,
+			ModifyIndex: 2,
+		},
+	}
+	other := &Node{
+		ID:              id,
+		Node:            node,
+		Datacenter:      datacenter,
+		Address:         address,
+		TaggedAddresses: make(map[string]string),
+		Meta:            make(map[string]string),
+		RaftIndex: RaftIndex{
+			CreateIndex: 1,
+			ModifyIndex: 3,
+		},
+	}
+	if !n.IsSame(other) {
+		t.Fatalf("should be equal, was %#v VS %#v", n, other)
+	}
+
+	other.ID = types.NodeID("")
+	if n.IsSame(other) {
+		t.Fatalf("should be different, was %#v VS %#v", n, other)
+	}
+	other.ID = id
+
+	other.Node = "other"
+	if n.IsSame(other) {
+		t.Fatalf("should be different, was %#v VS %#v", n, other)
+	}
+	other.Node = node
+
+	other.Datacenter = "dcX"
+	if n.IsSame(other) {
+		t.Fatalf("should be different, was %#v VS %#v", n, other)
+	}
+	other.Datacenter = datacenter
+
+	other.Address = "127.0.0.1"
+	if n.IsSame(other) {
+		t.Fatalf("should be different, was %#v VS %#v", n, other)
+	}
+	other.Address = address
+
+	other.TaggedAddresses = map[string]string{
+		"my": "address",
+	}
+	if n.IsSame(other) {
+		t.Fatalf("should be different, was %#v VS %#v", n, other)
+	}
+	other.TaggedAddresses = map[string]string{}
+	other.Meta = map[string]string{
+		"my": "meta",
+	}
+	if n.IsSame(other) {
+		t.Fatalf("should be different, was %#v VS %#v", n, other)
+	}
+	other.Meta = map[string]string{}
+	if !n.IsSame(other) {
+		t.Fatalf("should be equal, was %#v VS %#v", n, other)
+	}
+}
+
```

Comment:
It would be awesome to rewrite this as a table-driven test so we can really easily flesh out new cases with just a minimum of copy paste. There are existing examples like `TestStructs_NodeService_IsSame` which uses a helper function pattern, but more modern code uses a more standard table test for this kind of thing which has value in consistency.

Also, for new tests and modifying old ones, we prefer to use `testify/require` over the standard go `if/t.Fatal` pattern as it ends up being much easier to read.

More importantly though there is no test here for `ServiceNode.IsSame` and missing a field from that could cause this patch to behave significantly incorrectly and updates to never be observed - I don't think the end-to-end tests here cover every possibly field changing (reasonably so assuming that IsSame is correct) so this seems critical.

I'm in two minds about requesting a change here as we could get this into 1.3 build if we land it today but I think table tests are a really important thing for maintainability of methods like `IsSame` where we'll have to remember to update it and the tests every time we add new fields etc.

We'll discuss internally whether we are happy to merge it as it is or add a test ourselves to get it into 1.3.0.

I'd also like to see the equivalent one for the ServiceNode.IsSame which I don't see here